src/inputfetcher.h

```diff
@@ -0,0 +1,246 @@
+#include <unordered_set>
+#include <vector>
+
+/**
+ * Input fetcher for fetching inputs from the CoinsDB and inserting
+ * into the CoinsTip.
+ *
+ * The main thread loops through the block and writes all input prevouts to a
+ * global vector. It then wakes all workers and starts working as well. Each
+ * thread assigns itself a range of outpoints from the shared vector, and
+ * fetches the coins from disk. The outpoint and coin pairs are written to a
+ * thread local vector of pairs. Once all outpoints are fetched, the main thread
+ * loops through all thread local vectors and writes the pairs to the cache.
+ */
+class InputFetcher
+{
+private:
+    //! Mutex to protect the inner state
+    Mutex m_mutex{};
+    //! Worker threads block on this when out of work
+    std::condition_variable m_worker_cv{};
+    //! Main thread blocks on this when out of work
+    std::condition_variable m_main_cv{};
+
+    /**
+     * The outpoints to be fetched from disk.
+     * This is written to on the main thread, then read from all worker
+     * threads only after the main thread is done writing. Hence, it doesn't
+     * need to be guarded by a lock.
+     */
+    std::vector<COutPoint> m_outpoints{};
+    /**
+     * The index of the last outpoint that is being fetched. Workers assign
+     * themselves a range of outpoints to fetch from m_outpoints. They will use
+     * this index as the end of their range, and then set this index to the
+     * beginning of the range they take for the next worker. Once it gets to
+     * zero, all outpoints have been assigned and the next worker will wait.
+     */
+    size_t m_last_outpoint_index GUARDED_BY(m_mutex){0};
+
+    //! The set of txids of the transactions in the current block being fetched.
+    std::unordered_set<Txid, SaltedTxidHasher> m_txids{};
+    //! The vector of thread local vectors of pairs to be written to the cache.
+    std::vector<std::vector<std::pair<COutPoint, Coin>>> m_pairs{};
+
+    /**
+     * Number of outpoint fetches that haven't completed yet.
+     * This includes outpoints that have already been assigned, but are still in
+     * the worker's own batches.
+     */
+    int32_t m_in_flight_outpoints_count GUARDED_BY(m_mutex){0};
+    //! The number of worker threads that are waiting on m_worker_cv
+    int32_t m_idle_worker_count GUARDED_BY(m_mutex){0};
+    //! The maximum number of outpoints to be assigned in one batch
+    const int32_t m_batch_size;
+    //! DB coins view to fetch from.
+    const CCoinsView* m_db{nullptr};
+    //! The cache to check if we already have this input.
+    const CCoinsViewCache* m_cache{nullptr};
+
+    std::vector<std::thread> m_worker_threads;
+    bool m_request_stop GUARDED_BY(m_mutex){false};
+
+    //! Internal function that does the fetching from disk.
+    void Loop(int32_t index, bool is_main_thread = false) noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        auto local_batch_size{0};
+        auto end_index{0};
```

Comment:
I think we should add exact types here to make sure calculations like `end_index - local_batch_size` can't underflow

Reply:
I rewrote this part, these are gone now.